pallets/migration/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_runtime_upgrade() -> Weight {
+			// TODO: Insert some mechanism to prevent the accidential execution of the same
+			// 	upgreade twice. Probably best, to let the runtime version be the key for the
+			// 	map of old upgrades and if it already exists, do not do the upgrade!
+			let max_per_block = T::BlockWeights::get().max_block;
+
+			// In order to safely unwrap below, we do this check. Although, one might argue
+			// a chain with a maximum weight of zero, might be useless.
+			if max_per_block == <Weight as Zero>::zero() {
+				// As we are not updating the storage, this will basically prevent the upgrade from
+				// running.
+				return Zero::zero();
+			}
+
+			// The number of blocks we are going to block the chain for the upgreade
+			//
+			// NOTE:
+			// For the reasons of missing a remainder we are adding 1 extra block at the end,
+			// so that the upgreade will definitely fit into the given blocks.
+			let period: T::BlockNumber = <T::WeightToBlockNumber>::convert(
+				T::Upgrades::upgrade_weight()
+					.checked_div(max_per_block)
+					.expect("Maximum block weight is not zero. qed."),
+			)
+			.saturating_add(One::one());
+
+			let now = <frame_system::Pallet<T>>::block_number();
+			// Insert start and period into storage for the `on_initialize`
+			<UpgradeStart<T>>::set(now);
+			<UpgradePeriod<T>>::set(period);
+
+			let (version, memo) = T::Upgrades::info();
+			<UpgradeInfo<T>>::insert(now, (version.clone(), Vec::from(memo.as_ref())));
+
+			Self::deposit_event(Event::RuntimeUpgradeStarted(period, version));
+
+			// return zero as `on_initialize` will take care of the upgrade
+			Zero::zero()
+		}
+
+		fn on_initialize(n: T::BlockNumber) -> frame_support::weights::Weight {
+			// You can not do an upgrade on the first block...
+			if n < Self::upgrade_start().saturating_add(Self::upgrade_period()) && n != Zero::zero()
+			{
+				let period = Self::upgrade_period();
+				let start = Self::upgrade_start();
+				let max_per_block = T::BlockWeights::get().max_block;
+
+				Self::upgrade(max_per_block);
+
+				// Indicate to the network, that the upgrade has been executed successfully
+				if (n == start.saturating_add(period) - One::one())
+					&& <UpgradeInfo<T>>::contains_key(start)
+				{
+					let (version, memo) = <UpgradeInfo<T>>::get(start)
+						.expect("The storage is populated with the current upgread. qed.");
+
+					// Remove the last StateInfo from storage
+					<UpgradeState<T>>::set(None);
+
+					Self::deposit_event(Event::RuntimeUpgradeFinished(version, memo));
+				}
+
+				max_per_block
+			} else {
+				Zero::zero()
+			}
+		}
+	}
+
+	#[pallet::type_value]
+	pub fn OnUpgradeStartEmpty<T: Config>() -> T::BlockNumber {
+		Zero::zero()
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn upgrade_start)]
+	/// The start block of a runtime upgrade
+	pub(super) type UpgradeStart<T: Config> =
```

Comment:
instead of all these storages, can we use a upgradeInfo to store all these storage values in a struct? doesn't make a lot difference but with storage values, we would have to kill them as soon as we want to reuse them

Reply:
You mean somethink like that? 

```rust
pub struct StorageInfo {
     pub upgrade_start: T::BlockNumber,
     pub upgrade_period: T::BlockNumber
```

Yes, I am not opposed to that. Maybe even allows us to reduce the DB accesses. 
For curiosity, what do you mean with 

> [...] we would have to kill them as soon as we want to reuse them